fix: resolve duplicate invocation of overloaded lifecycle methods with inheritance

tangyang9464 opened the pull request:
fix: [3849](https://github.com/mapstruct/mapstruct/issues/3849)

Newly added `LifecycleOverloadDeduplicateSelector`, which uses multiple features such as **class name** and **method name** as keys to group methods.

And then selects the best method within the group by comparing the inheritance distance of parameters, 
to achieve deduplication between similar overloaded methods. 

The parameters involved in the inheritance distance calculation are not limited to `@MappingTarget`, but all input parameters of a group of candidate methods are traversed and calculated in sequence until a parameter is found with the shortest distance, then the traversal is terminated. (Avoid the situation where the **@MappingTarget** parameter type is the same but other parameters are different).

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java:
```diff
@@ -0,0 +1,131 @@
+        for ( ParameterBinding binding : method.getParameterBindings() ) {
+            key.add( binding.getType() );
+            key.add( binding.getVariableName() );
+            key.add( binding.isMappingTarget() );
+            key.add( binding.isMappingContext() );
+            key.add( binding.isTargetType() );
+            key.add( binding.isSourcePropertyName() );
+            key.add( binding.isTargetPropertyName() );
```
Why should these elements be part of the signature?

tangyang9464 replied:
This is not necessary.  The type and variable name are sufficient to determine the parameter signature.

filiphr replied:
Cool thanks, Shall we remove those then? And just so that I understand properly, when you say variable name, it is the name of something in the mapping method, not the name of a parameter in the lifecycle method, right?

tangyang9464 replied:
> Cool thanks, Shall we remove those then? And just so that I understand properly, when you say variable name, it is the name of something in the mapping method, not the name of a parameter in the lifecycle method, right?
Of course.

> when you say variable name, it is the name of something in the mapping method, not the name of a parameter in the lifecycle method, right?
Yes.

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java:
```diff
@@ -0,0 +1,131 @@
+        if ( !context.getSelectionCriteria().isLifecycleCallbackRequired() ) {
+            return methods;
+        }
+        List<List<SelectedMethod<T>>> methodSignatureGroups = new ArrayList<>(
+            methods.stream()
+                .collect( Collectors.groupingBy(
+                    LifecycleOverloadDeduplicateSelector::buildSignatureKey,
+                    LinkedHashMap::new,
+                    Collectors.toList()
+                ) )
+                .values()
```
Let's try to avoid creating a lot of objects. There is no need to create a new list for the values. You can just get the values (`Collection<List<SelectedMethod<T>>>`) and iterate over it

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java:
```diff
@@ -0,0 +1,131 @@
+                ) )
+                .values()
+        );
+        List<SelectedMethod<T>> deduplicatedMethods = new ArrayList<>();
```
Let's initialize the list with a size that is the `methods.size()`

filiphr commented on processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java:
```diff
@@ -0,0 +1,131 @@
+    @Override
+    public <T extends Method> List<SelectedMethod<T>> getMatchingMethods(List<SelectedMethod<T>> methods,
+                                                                         SelectionContext context) {
+        if ( !context.getSelectionCriteria().isLifecycleCallbackRequired() ) {
```
Let's add `methods.size() <= 1` as an additional short circuit

filiphr reviewed (changes requested):
Thanks @tangyang9464, the solution looks good to me. I've left some comments, which are mainly for performance optimization, as this can get called a lot, especially if you have a lot of lifecycle methods.

Additionally, I think that it'll be good if we can add some tests with generics. I am not sure whether everything would fine then.

Also, can we perhaps add some compiler flag so that people can disable this particular selector in case it causes them performance problems and / or something isn't working properly, just in case. I think we can pass the `Options` to the `MethodsSelectors` and then check whether we should include this new one or not. What do you think about that?

filiphr commented:
btw, the build is failing due to the javadoc

```
Error:  /home/runner/work/mapstruct/mapstruct/processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java:35: error: unknown tag: AfterMapping
Error:   * @AfterMapping default void afterMapping(Parent source, @MappingTarget ParentDto target) { ... }
Error:     ^
Error:  /home/runner/work/mapstruct/mapstruct/processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java:36: error: unknown tag: AfterMapping
Error:   * @AfterMapping default void afterMapping(Parent source, @MappingTarget ChildDto target) { ... }
Error:     ^
Error:  /home/runner/work/mapstruct/mapstruct/processor/src/main/java/org/mapstruct/ap/internal/model/source/selector/LifecycleOverloadDeduplicateSelector.java:37: error: unexpected end tag: </pre>
Error:   * </pre>
```

tangyang9464 commented:
I'll modify based on the comments and fix CICD later

tangyang9464 commented:
@filiphr plz review again.  I did the following:
1. Added compile args to disable the selectors and add related tests
2. Added generic tests
3. Optimized the code to improve performance based on your comments

tangyang9464 commented:
@filiphr any viewpoint？

filiphr commented:
Thanks for the changes @tangyang9464. During the week it's a bit more difficult for me to check these stuff out. I'll review it soon and merge it
